File: src/apps/desktop/src/api/acp_client_api.rs
//! ACP client API

use crate::api::app_state::AppState;
use crate::api::session_storage_path::desktop_effective_session_storage_path;
use bitfun_acp::client::{
    AcpClientInfo, AcpClientPermissionResponse, AcpClientRequirementProbe, AcpClientStreamEvent,
    AcpSessionOptions, CreateAcpFlowSessionRecordResponse, SetAcpSessionModelRequest,
    SubmitAcpPermissionResponseRequest,
};
use serde::{Deserialize, Serialize};
use tauri::{AppHandle, Emitter, State};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AcpClientIdRequest {
    pub client_id: String,
    #[serde(default)]
    pub remote_connection_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateAcpFlowSessionRequest {
    pub client_id: String,
    #[serde(default)]
    pub session_name: Option<String>,
    pub workspace_path: String,
    #[serde(default)]
    pub remote_connection_id: Option<String>,
    #[serde(default)]
    pub remote_ssh_host: Option<String>,
}

pub type CreateAcpFlowSessionResponse = CreateAcpFlowSessionRecordResponse;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StartAcpDialogTurnRequest {
    pub session_id: String,
    pub client_id: String,
    pub user_input: String,
    #[serde(default)]
    pub original_user_input: Option<String>,
    pub turn_id: String,
    #[serde(default)]
    pub workspace_path: Option<String>,
    #[serde(default)]
    pub remote_connection_id: Option<String>,
    #[serde(default)]
    pub remote_ssh_host: Option<String>,
    #[serde(default)]
    pub timeout_seconds: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CancelAcpDialogTurnRequest {
    pub session_id: String,
    pub client_id: String,
    #[serde(default)]
    pub workspace_path: Option<String>,
    #[serde(default)]
    pub remote_connection_id: Option<String>,
    #[serde(default)]
    pub remote_ssh_host: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GetAcpSessionOptionsRequest {
    pub session_id: String,
    pub client_id: String,
    #[serde(default)]
    pub workspace_path: Option<String>,
    #[serde(default)]
    pub remote_connection_id: Option<String>,
    #[serde(default)]
    pub remote_ssh_host: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ProbeAcpClientRequirementsRequest {
    #[serde(default)]
    pub remote_connection_id: Option<String>,
    #[serde(default)]
    pub force_refresh: bool,
}

#[tauri::command]
pub async fn initialize_acp_clients(state: State<'_, AppState>) -> Result<(), String> {
    let service = state
        .acp_client_service
        .as_ref()
        .ok_or_else(|| "ACP client service not initialized".to_string())?;
    service.initialize_all().await.map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_acp_clients(state: State<'_, AppState>) -> Result<Vec<AcpClientInfo>, String> {
    let service = state
        .acp_client_service
        .as_ref()
        .ok_or_else(|| "ACP client service not initialized".to_string())?;
    service.list_clients().await.map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn probe_acp_client_requirements(
    state: State<'_, AppState>,
    request: ProbeAcpClientRequirementsRequest,
) -> Result<Vec<AcpClientRequirementProbe>, String> {
    let service = state
        .acp_client_service
        .as_ref()
        .ok_or_else(|| "ACP client service not initialized".to_string())?;
    service
        .probe_client_requirements(
            request.remote_connection_id.as_deref(),
            request.force_refresh,
        )
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn predownload_acp_client_adapter(
    state: State<'_, AppState>,
    request: AcpClientIdRequest,
) -> Result<(), String> {
    let service = state
        .acp_client_service
        .as_ref()
        .ok_or_else(|| "ACP client service not initialized".to_string())?;
    service
        .predownload_client_adapter(&request.client_id)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn install_acp_client_cli(
    state: State<'_, AppState>,
    request: AcpClientIdRequest,
) -> Result<(), String> {
    let service = state
        .acp_client_service
        .as_ref()
        .ok_or_else(|| "ACP client service not initialized".to_string())?;
    service
        .install_client_cli(&request.client_id, request.remote_connection_id.as_deref())
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn create_acp_flow_session(
    state: State<'_, AppState>,
    app_handle: AppHandle,
    request: CreateAcpFlowSessionRequest,
) -> Result<CreateAcpFlowSessionResponse, String> {
    let service = state
        .acp_client_service
        .as_ref()
        .ok_or_else(|| "ACP client service not initialized".to_string())?;

    let session_storage_path = desktop_effective_session_storage_path(
        &state,
        &request.workspace_path,
        request.remote_connection_id.as_deref(),
        request.remote_ssh_host.as_deref(),
    )
    .await;
    let response = service
        .create_flow_session_record(
            &session_storage_path,
            &request.workspace_path,
            &request.client_id,
            request.session_name,
        )
        .await
        .map_err(|e| e.to_string())?;
    if let Err(error) = service
        .start_client_for_session(
            &request.client_id,
            &response.session_id,
            Some(&request.workspace_path),
            request.remote_connection_id.as_deref(),
        )
        .await
    {
        if let Err(cleanup_error) = service
            .delete_flow_session_record(&session_storage_path, &response.session_id)
            .await
        {
            log::warn!(
                "Failed to delete ACP session record after client start failure: session_id={}, error={}",
                response.session_id,
                cleanup_error
            );
        }
        return Err(error.to_string());
    }

    let _ = app_handle.emit(
        "agentic://session-created",
        serde_json::json!({
            "sessionId": response.session_id.clone(),
            "sessionName": response.session_name.clone(),
            "agentType": response.agent_type.clone(),
            "workspacePath": request.workspace_path,
            "remoteConnectionId": request.remote_connection_id,
            "remoteSshHost": request.remote_ssh_host,
        }),
    );

    Ok(response)
}

#[tauri::command]
pub async fn start_acp_dialog_turn(
    state: State<'_, AppState>,
    app_handle: AppHandle,
    request: StartAcpDialogTurnRequest,
) -> Result<(), String> {
    let service = state
        .acp_client_service
        .as_ref()
        .ok_or_else(|| "ACP client service not initialized".to_string())?
        .clone();

    let session_id = request.session_id.clone();
    let turn_id = request.turn_id.clone();
    let user_input = request.user_input.clone();
    let original_user_input = request
        .original_user_input
        .clone()
        .unwrap_or_else(|| request.user_input.clone());
    let session_storage_path = match request.workspace_path.as_deref() {
        Some(workspace_path) => Some(
            desktop_effective_session_storage_path(
                &state,
                workspace_path,
                request.remote_connection_id.as_deref(),
                request.remote_ssh_host.as_deref(),
            )
            .await,
        ),
        None => None,
    };

    app_handle
        .emit(
            "agentic://dialog-turn-started",
            serde_json::json!({
                "sessionId": session_id,
                "turnId": turn_id,
                "turnIndex": null,
                "userInput": user_input,
                "originalUserInput": original_user_input,
                "userMessageMetadata": null,
                "subagentParentInfo": null,
            }),
        )
        .map_err(|e| e.to_string())?;
    tokio::spawn(async move {
        let mut current_round_id: Option<String> = None;
        let result = service
            .prompt_agent_stream(
                &request.client_id,
                request.user_input,
                request.workspace_path,
                request.remote_connection_id,
                request.session_id.clone(),
                session_storage_path,
                request.timeout_seconds,
                |event| {
                    match event {
                        AcpClientStreamEvent::ModelRoundStarted {
                            round_id,
                            round_index,
                            disable_explore_grouping,
                        } => {
                            current_round_id = Some(round_id.clone());
                            app_handle
                                .emit(
                                    "agentic://model-round-started",
                                    serde_json::json!({
                                        "sessionId": request.session_id,
                                        "turnId": request.turn_id,
                                        "roundId": round_id,
                                        "roundIndex": round_index,
                                        "renderHints": {
                                            "disableExploreGrouping": disable_explore_grouping,
                                        },
                                        "subagentParentInfo": null,
                                    }),
                                )
                                .map_err(|e| {
                                    bitfun_core::util::errors::BitFunError::service(e.to_string())
                                })?;
                        }
                        AcpClientStreamEvent::AgentText(text) => {
                            let round_id = current_round_id.clone().ok_or_else(|| {
                                bitfun_core::util::errors::BitFunError::service(
                                    "ACP text arrived before model round start".to_string(),
                                )
                            })?;
                            app_handle
                                .emit(
                                    "agentic://text-chunk",
                                    serde_json::json!({
                                        "sessionId": request.session_id,
                                        "turnId": request.turn_id,
                                        "roundId": round_id,
                                        "text": text,
                                        "subagentParentInfo": null,
                                    }),
                                )
                                .map_err(|e| {
                                    bitfun_core::util::errors::BitFunError::service(e.to_string())
                                })?;
                        }
                        AcpClientStreamEvent::AgentThought(text) => {
                            let round_id = current_round_id.clone().ok_or_else(|| {
                                bitfun_core::util::errors::BitFunError::service(
                                    "ACP thought arrived before model round start".to_string(),
                                )
                            })?;
                            app_handle
                                .emit(
                                    "agentic://text-chunk",
                                    serde_json::json!({
                                        "sessionId": request.session_id,
                                        "turnId": request.turn_id,
                                        "roundId": round_id,
                                        "text": text,
                                        "contentType": "thinking",
                                        "isThinkingEnd": false,
                                        "subagentParentInfo": null,
                                    }),
                                )
                                .map_err(|e| {
                                    bitfun_core::util::errors::BitFunError::service(e.to_string())
                                })?;
                        }
                        AcpClientStreamEvent::ToolEvent(tool_event) => {
                            app_handle
                                .emit(
                                    "agentic://tool-event",
                                    serde_json::json!({
                                        "sessionId": request.session_id,
                                        "turnId": request.turn_id,
                                        "toolEvent": tool_event,
                                        "subagentParentInfo": null,
                                    }),
                                )
                                .map_err(|e| {
                                    bitfun_core::util::errors::BitFunError::service(e.to_string())
                                })?;
                        }
                        AcpClientStreamEvent::ContextUsageUpdated(usage) => {
                            app_handle
                                .emit(
                                    "agentic://acp-context-usage-updated",
                                    serde_json::json!({
                                        "sessionId": request.session_id,
                                        "turnId": request.turn_id,
                                        "clientId": request.client_id,
                                        "used": usage.used,
                                        "size": usage.size,
                                        "cost": usage.cost,
                                        "subagentParentInfo": null,
                                    }),
                                )
                                .map_err(|e| {
                                    bitfun_core::util::errors::BitFunError::service(e.to_string())
                                })?;
                        }
                        AcpClientStreamEvent::Completed => {
                            app_handle
                                .emit(
                                    "agentic://dialog-turn-completed",
                                    serde_json::json!({
                                        "sessionId": request.session_id,
                                        "turnId": request.turn_id,
                                        "subagentParentInfo": null,
                                        "partialRecoveryReason": null,
                                    }),
                                )
                                .map_err(|e| {
                                    bitfun_core::util::errors::BitFunError::service(e.to_string())
                                })?;
                        }
                        AcpClientStreamEvent::Cancelled => {
                            app_handle
                                .emit(
                                    "agentic://dialog-turn-cancelled",
                                    serde_json::json!({
                                        "sessionId": request.session_id,
                                        "turnId": request.turn_id,
                                        "subagentParentInfo": null,
                                    }),
                                )
                                .map_err(|e| {
                                    bitfun_core::util::errors::BitFunError::service(e.to_string())
                                })?;
                        }
                    }
                    Ok(())
                },
            )
            .await;

        if let Err(error) = result {
            let _ = app_handle.emit(
                "agentic://dialog-turn-failed",
                serde_json::json!({
                    "sessionId": request.session_id,
                    "turnId": request.turn_id,
                    "error": error.to_string(),
                    "errorCategory": null,
                    "errorDetail": null,
                    "subagentParentInfo": null,
                }),
            );
        }
    });

    Ok(())
}

#[tauri::command]
pub async fn cancel_acp_dialog_turn(
    state: State<'_, AppState>,
    request: CancelAcpDialogTurnRequest,
) -> Result<(), String> {
    let service = state
        .acp_client_service
        .as_ref()
        .ok_or_else(|| "ACP client service not initialized".to_string())?;
    service
        .cancel_agent_session(
            &request.client_id,
            request.workspace_path,
            request.session_id,
        )
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_acp_session_options(
    state: State<'_, AppState>,
    request: GetAcpSessionOptionsRequest,
) -> Result<AcpSessionOptions, String> {
    let service = state
        .acp_client_service
        .as_ref()
        .ok_or_else(|| "ACP client service not initialized".to_string())?;
    let session_storage_path = match request.workspace_path.as_deref() {
        Some(workspace_path) => Some(
            desktop_effective_session_storage_path(
                &state,
                workspace_path,
                request.remote_connection_id.as_deref(),
                request.remote_ssh_host.as_deref(),
            )
            .await,
        ),
        None => None,
    };
    service
        .get_session_options(
            &request.client_id,
            request.workspace_path,
            request.remote_connection_id,
            session_storage_path,
            request.session_id,
        )
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn set_acp_session_model(
    state: State<'_, AppState>,
    request: SetAcpSessionModelRequest,
) -> Result<AcpSessionOptions, String> {
    let service = state
        .acp_client_service
        .as_ref()
        .ok_or_else(|| "ACP client service not initialized".to_string())?;
    let session_storage_path = match request.workspace_path.as_deref() {
        Some(workspace_path) => Some(
            desktop_effective_session_storage_path(
                &state,
                workspace_path,
                request.remote_connection_id.as_deref(),
                request.remote_ssh_host.as_deref(),
            )
            .await,
        ),
        None => None,
    };
    service
        .set_session_model(request, session_storage_path)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn stop_acp_client(
    state: State<'_, AppState>,
    request: AcpClientIdRequest,
) -> Result<(), String> {
    let service = state
        .acp_client_service
        .as_ref()
        .ok_or_else(|| "ACP client service not initialized".to_string())?;
    service
        .stop_client(&request.client_id)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn load_acp_json_config(state: State<'_, AppState>) -> Result<String, String> {
    let service = state
        .acp_client_service
        .as_ref()
        .ok_or_else(|| "ACP client service not initialized".to_string())?;
    service.load_json_config().await.map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn save_acp_json_config(
    state: State<'_, AppState>,
    json_config: String,
) -> Result<(), String> {
    let service = state
        .acp_client_service
        .as_ref()
        .ok_or_else(|| "ACP client service not initialized".to_string())?;
    service
        .save_json_config(&json_config)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn submit_acp_permission_response(
    state: State<'_, AppState>,
    request: SubmitAcpPermissionResponseRequest,
) -> Result<AcpClientPermissionResponse, String> {
    let service = state
        .acp_client_service
        .as_ref()
        .ok_or_else(|| "ACP client service not initialized".to_string())?;
    service
        .submit_permission_response(request)
        .await
        .map_err(|e| e.to_string())
}

File: src/web-ui/src/infrastructure/api/service-api/ACPClientAPI.ts
import { api } from './ApiClient';
import type { ImageContextData as ImageInputContextData } from './ImageContextTypes';

export type AcpClientPermissionMode = 'ask' | 'allow_once' | 'reject_once';
export type AcpClientStatus = 'configured' | 'starting' | 'running' | 'stopped' | 'failed';

export interface AcpClientInfo {
  id: string;
  name: string;
  command: string;
  args: string[];
  enabled: boolean;
  readonly: boolean;
  permissionMode: AcpClientPermissionMode;
  status: AcpClientStatus;
  toolName: string;
  sessionCount: number;
}

export interface AcpRequirementProbeItem {
  name: string;
  installed: boolean;
  version?: string;
  path?: string;
  error?: string;
}

export interface AcpClientRequirementProbe {
  id: string;
  tool: AcpRequirementProbeItem;
  adapter?: AcpRequirementProbeItem;
  runnable: boolean;
  notes: string[];
}

export interface AcpClientIdRequest {
  clientId: string;
  remoteConnectionId?: string;
}

export interface CreateAcpFlowSessionRequest {
  clientId: string;
  sessionName?: string;
  workspacePath: string;
  remoteConnectionId?: string;
  remoteSshHost?: string;
}

export interface CreateAcpFlowSessionResponse {
  sessionId: string;
  sessionName: string;
  agentType: string;
}

export interface StartAcpDialogTurnRequest {
  sessionId: string;
  clientId: string;
  userInput: string;
  originalUserInput?: string;
  turnId: string;
  workspacePath?: string;
  remoteConnectionId?: string;
  remoteSshHost?: string;
  timeoutSeconds?: number;
  imageContexts?: ImageInputContextData[];
  userMessageMetadata?: Record<string, unknown>;
}

export interface CancelAcpDialogTurnRequest {
  sessionId: string;
  clientId: string;
  workspacePath?: string;
  remoteConnectionId?: string;
  remoteSshHost?: string;
}

export interface GetAcpSessionOptionsRequest {
  sessionId: string;
  clientId: string;
  workspacePath?: string;
  remoteConnectionId?: string;
  remoteSshHost?: string;
}

export interface SetAcpSessionModelRequest {
  sessionId: string;
  clientId: string;
  workspacePath?: string;
  remoteConnectionId?: string;
  remoteSshHost?: string;
  modelId: string;
}

export interface AcpSessionModelOption {
  id: string;
  name: string;
  description?: string;
}

export interface AcpContextUsage {
  used: number;
  size: number;
  cost?: {
    amount: number;
    currency: string;
  };
}

export interface AcpSessionOptions {
  currentModelId?: string;
  availableModels: AcpSessionModelOption[];
  modelConfigId?: string;
  contextUsage?: AcpContextUsage;
}

export interface SubmitAcpPermissionResponseRequest {
  permissionId: string;
  approve: boolean;
  optionId?: string;
}

export interface AcpPermissionOption {
  optionId: string;
  name: string;
  kind: 'allow_once' | 'allow_always' | 'reject_once' | 'reject_always';
}

export interface AcpPermissionRequestEvent {
  permissionId: string;
  sessionId: string;
  toolCall?: {
    toolCallId?: string;
    title?: string;
    rawInput?: unknown;
    content?: unknown;
  };
  options?: AcpPermissionOption[];
}

const LOCAL_REQUIREMENT_CACHE_KEY = '__local__';
const requirementProbeCache = new Map<string, AcpClientRequirementProbe[]>();
const requirementProbeInFlight = new Map<string, Promise<AcpClientRequirementProbe[]>>();

export class ACPClientAPI {
  private static invalidateRequirementProbeCache(): void {
    requirementProbeCache.clear();
    requirementProbeInFlight.clear();
  }

  static async initializeClients(): Promise<void> {
    await api.invoke('initialize_acp_clients');
    ACPClientAPI.invalidateRequirementProbeCache();
    window.dispatchEvent(new Event('bitfun:acp-clients-changed'));
  }

  static async getClients(): Promise<AcpClientInfo[]> {
    return api.invoke('get_acp_clients');
  }

  static async probeClientRequirements(
    options: { force?: boolean; remoteConnectionId?: string } = {}
  ): Promise<AcpClientRequirementProbe[]> {
    const cacheKey = options.remoteConnectionId || LOCAL_REQUIREMENT_CACHE_KEY;
    if (!options.force && requirementProbeCache.has(cacheKey)) {
      return requirementProbeCache.get(cacheKey) ?? [];
    }
    if (!options.force && requirementProbeInFlight.has(cacheKey)) {
      return requirementProbeInFlight.get(cacheKey)!;
    }

    const request = options.remoteConnectionId
      ? { remoteConnectionId: options.remoteConnectionId, forceRefresh: options.force === true }
      : {};

    const inFlight = api.invoke<AcpClientRequirementProbe[]>('probe_acp_client_requirements', { request })
      .then((probes) => {
        requirementProbeCache.set(cacheKey, probes);
        window.dispatchEvent(new Event('bitfun:acp-requirements-changed'));
        return probes;
      })
      .finally(() => {
        requirementProbeInFlight.delete(cacheKey);
      });

    requirementProbeInFlight.set(cacheKey, inFlight);
    return inFlight;
  }

  static async predownloadClientAdapter(request: AcpClientIdRequest): Promise<void> {
    await api.invoke('predownload_acp_client_adapter', { request });
    ACPClientAPI.invalidateRequirementProbeCache();
    window.dispatchEvent(new Event('bitfun:acp-requirements-changed'));
  }

  static async installClientCli(request: AcpClientIdRequest): Promise<void> {
    await api.invoke('install_acp_client_cli', { request });
    ACPClientAPI.invalidateRequirementProbeCache();
    window.dispatchEvent(new Event('bitfun:acp-requirements-changed'));
  }

  static async stopClient(request: AcpClientIdRequest): Promise<void> {
    await api.invoke('stop_acp_client', { request });
    window.dispatchEvent(new Event('bitfun:acp-clients-changed'));
  }

  static async loadJsonConfig(): Promise<string> {
    return api.invoke('load_acp_json_config');
  }

  static async saveJsonConfig(jsonConfig: string): Promise<void> {
    await api.invoke('save_acp_json_config', { jsonConfig });
    ACPClientAPI.invalidateRequirementProbeCache();
    window.dispatchEvent(new Event('bitfun:acp-clients-changed'));
  }

  static async submitPermissionResponse(
    request: SubmitAcpPermissionResponseRequest
  ): Promise<void> {
    return api.invoke('submit_acp_permission_response', { request });
  }

  static async createFlowSession(
    request: CreateAcpFlowSessionRequest
  ): Promise<CreateAcpFlowSessionResponse> {
    const response = await api.invoke<CreateAcpFlowSessionResponse>('create_acp_flow_session', { request });
    window.dispatchEvent(new Event('bitfun:acp-clients-changed'));
    return response;
  }

  static async startDialogTurn(request: StartAcpDialogTurnRequest): Promise<void> {
    return api.invoke('start_acp_dialog_turn', { request });
  }

  static async cancelDialogTurn(request: CancelAcpDialogTurnRequest): Promise<void> {
    return api.invoke('cancel_acp_dialog_turn', { request });
  }

  static async getSessionOptions(
    request: GetAcpSessionOptionsRequest
  ): Promise<AcpSessionOptions> {
    return api.invoke('get_acp_session_options', { request });
  }

  static async setSessionModel(
    request: SetAcpSessionModelRequest
  ): Promise<AcpSessionOptions> {
    return api.invoke('set_acp_session_model', { request });
  }
}

export default ACPClientAPI;
